Adding L-moments fit

efvik opened the pull request:
Here is one solution for what was discussed in e.g. #92 . Let me know what you think, @georgebv or others.

Changes:
 * Class `Distribution`, in `distribution.py`, now generally models any scipy continuous distribution, not only MLE. This is controlled with "method" parameter in initialization. For now, I've suggested "MOM" and "LMOM" for method of moments and L-moments.
 * New class `ScipyModel` inherets from `AbstractModelBaseClass`, and is a generalization of the old class `MLE`
 * New classes `MLE(ScipyModel)`,`MOM(ScipyModel)`,`LMOM(ScipyModel)`, represent each fitting method.

L-Moment fit relies on the [lmoments3](https://github.com/OpenHydrology/lmoments3) package. In my experience it is more robust than MLE, and can be around 10x faster, which is useful on very large datasets.

<details>
  <summary> Comparisons </summary>

<img width="849" height="1223" alt="image" src="https://github.com/user-attachments/assets/9d3e2f19-ffc9-4260-97c6-7e76ad756330" />

<img width="865" height="1223" alt="image" src="https://github.com/user-attachments/assets/6f6f0496-b833-4cda-add3-cd6614034815" />

</details>

georgebv commented on src/pyextremes/models/distribution.py:
```diff
@@ -53,11 +61,39 @@ def __init__(
+                }
+                if distribution in scipy_to_lmom.keys():
+                    distribution = scipy_to_lmom[distribution]
+                self.distribution = getattr(lmoments3.distr, distribution)
```
I suggest keeping distribution an instance of scipy rv_continuour, but instead adding another attrbiute here for type safety.

efvik replied:
I'm not sure what you mean. Here, `self.distribution` becomes an instance of `scipy.rv_continous`, assuming `distribution` parameter is a string. Did I overlook a potential issue?

georgebv commented on src/pyextremes/models/distribution.py:
```diff
@@ -53,11 +61,39 @@ def __init__(
+                self.distribution = getattr(lmoments3.distr, distribution)
+
+                for k in kwargs: 
+                    if k != "floc": 
```
Should this be `==`?

efvik replied:
Well, consider this 
is related to the mess down here, :
As far as I can see, there is no way to fix parameters in lmoments3. That is why I raise an error if there is any kwarg (since the kwargs to the distribution class are any fixed parameters like floc, fscale, ...).
```python
for k in kwargs: 
    if k != "floc": 
        raise ValueError("L-moments does not allow fixed parameters.")
```
However, when fitting distribution to Peak-over-Threshold extremes, you set the location parameter of the distribution to be fixed and equal to the threshold, which doesn't work with L-moments. But as I understand it, the equivalent method is to simply subtract the threshold from all the extremes. This is what I did here, in `Distribution.fit()`:
```
if self.method == 'LMOM':
    floc = self.fixed_parameters.get("floc",0.)
    parameters = self.distribution.lmom_fit(data=data-floc)
    parameters = list(parameters.values())
    if "floc" in self.fixed_parameters:
        self.fixed_parameters["floc"] += parameters[-2]
        self._fixed_parameters["loc"] += parameters[-2]
        parameters[-2] = self.fixed_parameters["floc"]
```
It's not very elegant, but it seems to work. I would appreciate if you have any comments on or potential issues with this method, as there might very well be some issues here I haven't considered.

georgebv replied:
Are you speaking about this? https://github.com/georgebv/pyextremes/blob/27c9afe724f0590da06986cb7483d1986b5acc75/src/pyextremes/eva.py#L1028-L1033

We can change this behavior for lmom.

georgebv reviewed (changes requested):
Thank you for the effort! I like the direction where this is going. The comments I have are mostly about backwards compatibility. This package has been out for a while and people have code that depends on specific API. It should pass all existing tests and we should add a few for the new models. I would also like the new package to be an optional dependency (installed like `pip install pyextremes[lmom]`).

Let me know if you need help with any parts of this.

efvik commented:
Hi, sorry for the delay. I was on sick leave for a while.

I think I have addressed most of the issues above in the new commits, but let me know if I missed anything.

1. I reverted the change to the mle_parameters attribute, and all existing tests should now be OK again. I'm not sure what you mean regarding adding another attribute - would it just duplicate the content of the current mle_parameters? I personally don't care if all distributions use the existing mle_parameters, even if the name is a little bit misleading. 

2. Do you have any other suggestions for the fitting methods? I have seen MOM in literature as an initialism of "Method of Moments" (although MM might be just as common, as used by scipy). In either case, LMOM is somewhat jarring next to MOM and MLE, since it suggests an initialism of "Linear Method of Moments". Perhaps changing to "MM" and "Lmoments" would be better, and avoid any confusion?

3. On a related note, I'm not sure if you want the method of moments to be included or not, or perhaps it can be included with a warning. In my experience it very often fails to converge when fitting only to extremes.

4. I can probably add some tests too, but do you have a suggestion for how to do this in a clean way with the lmoments3 package being optional?

georgebv commented:
@efvik thank you for the update! I like your suggestion of naming them `MOM` (I think that's how most people abbreviate method-of-moments) and `Lmoments` to avoid confusion. If you think it's worth it to include only `Lmoments` that would be fine with me. For tests I think the approach would be creating new dependency group `test-lmoments`, xfailing tests that test lmoments without that dep, and then running regular test and test-lmoments tests in a matrix in CI. I can handle that configuration in a separate PR, but could you please for now add lmoments to the regular `test` dependency group and write some tests for it?

efvik commented:
@georgebv I changed the name to `Lmoments`, and added a file `test_model_lmoments.py` (mostly a copy of `test_model_mle.py`). I don't really know how to account for the various scenarios with missing dependencies, but let me know if it needs to be done differently.

I suppose I don't see a reason to not include method of moments. It's probably not a good choice in 99% of cases, but since pyextremes currently allows us to choose any extremes (via `set_extremes`) I think it makes sense to also allow all types of fitting methods as options. I know method of moments is a common choice for the "initial distribution approach" (fitting to all data). 

PS: Would you mind commenting on #104 ?

georgebv reviewed (commented):
@efvik thank you and apologies for slow response time (also responded on the https://github.com/georgebv/pyextremes/discussions/104 thread, it's an interesting topic).

I left a few comments, but overall I think we should be ready to ship (you can ignore the codecov complaining about missing coverage for import lines). Could you please write a few test cases for MOM as well?

efvik commented on src/pyextremes/models/distribution.py:
```diff
@@ -128,23 +172,30 @@ def fit(self, data: np.ndarray) -> dict:
-        # Calculate full MLE of distribution parameters
-        full_mle = self.distribution.fit(data=data, **self.fixed_parameters)
 
+        # Calculate full distribution parameters
+        if self.method == 'MLE':
+            parameters = self.distribution.fit(data=data, **self.fixed_parameters, method='mle')
+        if self.method == 'MOM':
+            parameters = self.distribution.fit(data=data, **self.fixed_parameters, method='mm')
+        if self.method == 'Lmoments':
+            parameters = self.distribution.lmom_fit(data=data)
+            parameters = list(parameters.values())
```
I don't think this can ever happen, since we check the method input on initialization, but I suppose there's no reason not to. Added.

efvik commented:
Thank you for the update here and for the comments on the discussion as well.

I added your suggested changes, and also created a tests file for the method of moments similar to the MLE and Lmoments tests. Method of moments does not seem to want to fit on genextreme and genpareto, so I had to skip those, but it seems to work for gumbel and exponential. As I said above, I think this method is mainly used for fitting to a full distribution of data and not just to the peaks/extremes, but I think it is nice to leave this choice for the user.

efvik commented:
Great! I see there are some linting errors, but I don't get them locally when running ruff check locally, so I don't know how to fix them.

Sure, I could help with maintaining. I don't have too much time available, but since the general level of activity on here seems quite low I think it should be more than manageable.
